docs/how-to/hip_porting_guide.rst

```diff
@@ -27,51 +57,81 @@ error code spaces:
+  translates it into an Abstract Syntax Tree, and generates the HIP source. For this,
+  ``hipify-clang`` needs to be able to actually compile the code, so the CUDA code needs
+  to be correct, and a CUDA install with all necessary headers must be provided.
+
+* :doc:`hipify-perl <hipify:how-to/hipify-perl>` uses pattern matching, to translate the
+  CUDA code to HIP. It does not require a working CUDA installation, and can also
+  convert CUDA code, that is not syntactically correct. It is therefore easier to
+  set up and use, but is not as powerful as ``hipfiy-clang``.
+
+Memory copy functions
+---------------------
+
+When copying memory, the CUDA driver includes the memory direction in the name of
+the API (``cuMemcpyHtoD``), while the CUDA runtime API provides a single memory
+copy API with a parameter that specifies the direction. It also supports a
+default direction where the runtime determines the direction automatically.
+
+HIP provides both versions, for example, :cpp:func:`hipMemcpyHtoD` as well as
+:cpp:func:`hipMemcpy`. The first version might be faster in some cases because
+it avoids any host overhead to detect the direction of the memory copy.
+
+Address spaces
+--------------
 
-:doc:`hipify-clang <hipify:how-to/hipify-clang>` is, as the name implies, a Clang-based
-tool, and actually parses the code, translates it into an Abstract Syntax Tree,
-from which it then generates the HIP source. For this, ``hipify-clang`` needs to
-be able to actually compile the code, so the CUDA code needs to be correct, and
-a CUDA install with all necessary headers must be provided.
+HIP-Clang defines a process-wide address space where
+the CPU and all devices allocate addresses from a single unified pool.
+This means addresses can be shared between contexts. Unlike CUDA, a new context
+does not create a new address space for the device.
 
-:doc:`hipify-perl <hipify:how-to/hipify-perl>` uses pattern matching, to translate the
-CUDA code to HIP. It does not require a working CUDA installation, and can also
-convert CUDA code, that is not syntactically correct. It is therefore easier to
-set up and use, but is not as powerful as ``hipify-clang``.
+Context stack behavior differences
+----------------------------------
 
-Scanning existing CUDA code to scope the porting effort
---------------------------------------------------------------------------------
+HIP-Clang creates a primary context when the HIP API is called. In CUDA
+driver API code, HIP-Clang creates a primary context while HIP/NVCC has an empty
+context stack. HIP-Clang pushes the primary context to the context stack when it
+is empty. This can lead to subtle differences in applications which mix the
+runtime and driver APIs. 
 
-The ``--examine`` option, supported by the clang and perl version, tells hipify
+Scanning CUDA source to scope the translation
+---------------------------------------------
+
+The ``--examine`` option, supported by the clang and perl versions, tells hipify
```

Comment:
The ``--examine`` option, supported by both HIPIFY tools, tells the hipify tool
to do a test-run